cpukit/score/src/schedulerstrongapa.c

```diff
@@ -3,339 +3,820 @@
+  Scheduler_strong_APA_Context *self;
+
+  self = _Scheduler_strong_APA_Get_self( context );
+  lowest_scheduled = NULL; /* To remove compiler warning */
+  max_priority_num = 0;    /* Max (Lowest) priority encountered so far */
+  filter_strong_node = _Scheduler_strong_APA_Node_downcast( filter_base );
+
+  /* lowest_scheduled is NULL if affinity of a node is 0 */
+  _Assert( !_Processor_mask_Is_zero( &filter_strong_node->Affinity ) );
+  cpu_max = _SMP_Get_processor_maximum();
+
+  for ( cpu_index = 0 ; cpu_index < cpu_max ; ++cpu_index ) {
+    /* Checks if the CPU is in the affinity set of filter_strong_node */
+    if ( _Processor_mask_Is_set( &filter_strong_node->Affinity, cpu_index ) ) {
+      Per_CPU_Control *cpu = _Per_CPU_Get_by_index( cpu_index );
+
+      if ( _Per_CPU_Is_processor_online( cpu ) ) {
+        curr_node = _Scheduler_strong_APA_Get_scheduled( self, cpu );
+        curr_priority = _Scheduler_Node_get_priority( curr_node );
+        curr_priority = SCHEDULER_PRIORITY_PURIFY( curr_priority );
+
+        if ( curr_priority > max_priority_num ) {
+          lowest_scheduled = curr_node;
+          max_priority_num = curr_priority;
+        }
+      }
+    }
+  }
+
+  _Assert( lowest_scheduled != NULL );
+  return lowest_scheduled;
 }
 
-static void _Scheduler_strong_APA_Do_update(
+static inline void _Scheduler_strong_APA_Extract_from_scheduled(
   Scheduler_Context *context,
-  Scheduler_Node *node_to_update,
-  Priority_Control new_priority
+  Scheduler_Node    *node_to_extract
 )
 {
-  Scheduler_strong_APA_Context *self =
-    _Scheduler_strong_APA_Get_self( context );
-  Scheduler_strong_APA_Node *node =
-    _Scheduler_strong_APA_Node_downcast( node_to_update );
-
-  _Scheduler_SMP_Node_update_priority( &node->Base, new_priority );
-  _Scheduler_priority_Ready_queue_update(
-    &node->Ready_queue,
-    SCHEDULER_PRIORITY_UNMAP( new_priority ),
-    &self->Bit_map,
-    &self->Ready[ 0 ]
-  );
+  Scheduler_strong_APA_Context *self;
+  Scheduler_strong_APA_Node    *node;
+
+  self = _Scheduler_strong_APA_Get_self( context );
+  node = _Scheduler_strong_APA_Node_downcast( node_to_extract );
+
+  _Scheduler_SMP_Extract_from_scheduled( &self->Base.Base, &node->Base.Base );
+  /* Not removing it from Ready since the node could go in the READY state */
 }
 
-static Scheduler_strong_APA_Context *
-_Scheduler_strong_APA_Get_context( const Scheduler_Control *scheduler )
+static inline void _Scheduler_strong_APA_Extract_from_ready(
+  Scheduler_Context *context,
+  Scheduler_Node    *node_to_extract
+)
 {
-  return (Scheduler_strong_APA_Context *) _Scheduler_Get_context( scheduler );
+  Scheduler_strong_APA_Context *self;
+  Scheduler_strong_APA_Node    *node;
+
+  self = _Scheduler_strong_APA_Get_self( context );
+  node = _Scheduler_strong_APA_Node_downcast( node_to_extract );
+
+  if( !_Chain_Is_empty( &self->Ready ) ) {
+    if( !_Chain_Is_node_off_chain( &node->Ready_node ) ) {
```

Comment:
Just curious, is there any way for us to be sure that the `node->Ready_node` is on the `self->Ready` chain? (I'm thinking not, but I'm not sure if there could be any possibility that may happen somehow, and whether it could cause a bug? Then, we need to consider what race conditions might cause it to happen.)

Reply:
Actually, we don't even need to check  if( !_Chain_Is_empty( &self->Ready ), it is now removed. 

The check: if( !_Chain_Is_node_off_chain( &node->Ready_node ) is enough because node->Ready_node only gets inserted into the self->Ready chain, there is no other chain that it gets inserted in. Again, about the race condition - can two cores in an SMP system access(modify) the Scheduler_strong_APA_Context at the same time?

Reply:
I'm not sure what the concurrency rules are for the SMP scheduler framework now. I haven't implemented a scheduler in it. :)

Reply:
So, now that we know that a scheduler lock on Strong APA scheduler needs to be acquired, and the self->Ready chain is the only place where node->Ready_node gets inserted, is this resolved?

Reply:
yes

Reply:
Thanks.